system/datatypes/mailinglist.php

```diff
@@ -770,6 +782,22 @@ private function createOrUpdateMailing(array $data) {
 		return $mailing;
 	}
 
+	private function testSendMailing($mailingId, $email) {
+		if (!$this->hasSender()) {
+			throw new \Exception(__('ml-no-sender'));
+		}
+
+		$senderName = $this->getDoc()->getAttribute('sender-name');
+		$senderEmail = $this->getDoc()->getAttribute('sender-email');
+		
+		/** @var HyphaDomElement $mailing */
+		$mailing = $this->xml->document()->getElementById($mailingId);
+		
+		$message = '<p><a href="' . $linkToMailing . '">' . __('ml-if-unreadable-use-link') . '</a></p>';
+		$message .= $mailing->getHtml();
+		$this->send($mailing->getAttribute('subject'), $message, [$email], $senderEmail, $senderName);
```

Comment:
Same here: This code is duplicated between test send and regular send, so it would be very easy now to update one and not the other, making the test send non-representative. I can imagine merging these two functions (with a test boolean argument), or perhaps moving the duplicate code that generates (and perhaps sends) each message into its own function.